### PR TITLE
[AST] Introduce bits to capture whether a declaration is exposed to Objective-C

### PR DESCRIPTION
**Pull request by DougGregor**
The common query `ValueDecl::isObjC()` was checking for the presence of `ObjCAttr`, but didn't distinguish between "this has been checked and we know it is correct" and "this has not been checked yet", which means we could be wrong in both directions: we might not realize that something is exposed to Objective-C before we checked that the value could be computed, and we might think something is exposed to Objective-C when it fact it would be ill-formed to do so.

Start detangling this by adding bits for "is exposed to Objective-C" and "did we compute whether this is exposed to Objective-C" to `ValueDecl`, and use that information for the `ValueDecl::isObjC()` query---not the presence of `ObjCAttr`. Stop creating implicit `ObjCAttr` nodes unless we need them to store additional information, and make sure that the `is exposed to Objective-C` bit is consistently (de-)serialized.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -740,6 +740,16 @@ void PrintAST::printAttributes(const Decl *D) {
     Options.ExcludeAttrList.push_back(DAK_Final);
   }
 
+  // If the declaration is implicitly @objc, print the attribute now.
+  if (Options.PrintImplicitAttrs) {
+    if (auto VD = dyn_cast<ValueDecl>(D)) {
+      if (VD->isObjC() && !VD->getAttrs().hasAttribute<ObjCAttr>()) {
+        Printer.printAttrName("@objc");
```
[the remark this answers is not in the document]
Comment (DougGregor) [reply lib/AST/ASTPrinter.cpp]: It suppressed the printing if there is a “real” @objc attribute on the decl

[the remark this answers is not in the document]
Comment (slavapestov) [reply lib/AST/ASTPrinter.cpp]: Oh, I totally missed that. Oops!

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Pull request merged**